### ISSUE TITLE
Question about Squiz.WhiteSpace.FunctionSpacing and nested functions

### ISSUE THREAD

**Issue by em92** [squizlabs/PHP_CodeSniffer#2627]
There is a project RSS-Bridge that uses PHP_CodeSniffer https://github.com/RSS-Bridge/rss-bridge/

Last commit was this: https://github.com/RSS-Bridge/rss-bridge/commit/50b234d89349669868f057d7c6884c2068f6a745 with passed lint tests https://travis-ci.org/RSS-Bridge/rss-bridge/builds/585741688

Recently new PR have failed lint results. Let take example: https://travis-ci.org/RSS-Bridge/rss-bridge/jobs/592299733

```
FILE: /home/travis/build/RSS-Bridge/rss-bridge/bridges/ReadComicsBridge.php

---------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

---------------------------------------------------------------------------

 20 | ERROR | [x] Expected 0 blank lines before function; 1 found

---------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

---------------------------------------------------------------------------

FILE: /home/travis/build/RSS-Bridge/rss-bridge/bridges/Releases3DSBridge.php

----------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

----------------------------------------------------------------------------

 12 | ERROR | [x] Expected 0 blank lines before function; 1 found

----------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

----------------------------------------------------------------------------

FILE: /home/travis/build/RSS-Bridge/rss-bridge/bridges/ThePirateBayBridge.php

-----------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

-----------------------------------------------------------------------------

 43 | ERROR | [x] Expected 0 blank lines before function; 1 found

-----------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

-----------------------------------------------------------------------------

FILE: /home/travis/build/RSS-Bridge/rss-bridge/bridges/JapanExpoBridge.php

--------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

--------------------------------------------------------------------------

 22 | ERROR | [x] Expected 0 blank lines before function; 1 found

--------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

--------------------------------------------------------------------------
```
These files are not modifed from last commit. All error lines are on nested functions.

Checked wiki https://github.com/squizlabs/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties#squizwhitespacefunctionspacing
It says that `spacingBeforeFirst` applies to function in a class, trait, interface, but NOT in function. For other cases it uses value of `spacing` param. Also it looks like it worked as I told 2 weeks ago, but does not work now. Is it a bug, or there is a breaking update?

Related issue: https://github.com/RSS-Bridge/rss-bridge/issues/1306

**Comment by gsherwood** [comment]
Version 3.5.0 of PHPCS included this change:
> Squiz.WhiteSpace.FunctionSpacing now applies beforeFirst and afterLast spacing rules to nested functions
-- Previously, these rules only applied to the first and last function in a class, interface, or trait
-- These rules now apply to functions nested in any statement block, including other functions and conditions

This is a fix for issue #2406. The sniff itself never considered nested functions because the Squiz standard has always banned them and they are rarely used by other codebases, so it was not reporting the error correctly.

However, I did forget to update the documentation. I will do this now.

Also, this rule not only applies to nested functions,  but any grouping structure. So even a statement group like this:
```php
{

function foo() {}

function bar() {}

}
```
Would apply those beforeFirst and afterLast rules. Essentially, anywhere you might have written out several function declarations in a row should be treated like a function group and have consistent rules applied. I'm not sure if you use that syntax, but it's something to be aware of.

I notice that the nested functions have since been converted into class methods, so I'm assume you don't require any additional changes in this sniff any more. Is that the case?

Either way, hopefully this explains why you were seeing new errors and why the change was considered a bug fix.

**Comment by gsherwood** [comment]
I've updated those docs to "when it is the first block of code inside a statement group or control structure" as this covers all the cases that the sniff is checking. This includes things like namespace declarations that use curly braces, if statements that check if a function is defined, etc.
